lib/IRGen/GenClangType.cpp

```diff
@@ -606,12 +611,16 @@ clang::CanQualType GenClangType::visitSILFunctionType(CanSILFunctionType type) {
                              paramTy.getArgumentType(IGM.getSILModule(), type));
     if (param.isNull())
       return clang::CanQualType();
+
     paramTypes.push_back(param);
+    extParamInfos.push_back(extParamInfo);
   }
   
   // Build the Clang function type.
-  clang::FunctionProtoType::ExtProtoInfo defaultEPI;
-  auto fnTy = clangCtx.getFunctionType(resultType, paramTypes, defaultEPI);
+  clang::FunctionProtoType::ExtProtoInfo extProtoInfo;
+  extProtoInfo.ExtParameterInfos = extParamInfos.begin();
```

Comment:
@slavapestov -- is `SmallVector` guaranteed to have contiguous storage? And is `clangCtx.getFunctionType` guaranteed to create a copy of this array?

Reply:
Yes to both of these.

Reply:
Do you know if these points are documented anywhere? I looked at the Doxygen and I couldn't find this. Also, is there some general guidelines at play here (e.g. "getFooType" always deep-copies everything in Clang) or is this a "yes, I know these particular points are true"?

Reply:
All AST nodes are supposed to be self-contained in both Clang and Swift (except inasmuch as they reference other AST nodes, which should themselves be self-contained).  Creating an AST node should always be deep-copying any necessary arrays into trailing storage or some other associated storage.

There might be exceptions for `TypeVariableType`s in Swift, they're a weird case.